### PR TITLE
feat: display the crate's properties as hyperlink on the terminal

### PR DESCRIPTION
**Pull request by wangziling**
image

image

### CODE REVIEW

File: src/ops/view.rs [hunk 1]
```diff
@@ -12,6 +15,150 @@ use cargo::{
 
 use super::style::{ERROR, HEADER, LITERAL, NOP, NOTE, WARN};
 
+#[allow(unused)]
+#[derive(PartialEq, Eq)]
+enum Hyperlink<'a> {
```
Comment (epage) [inline src/ops/view.rs:20]: Through `GlobalContext` you can get a `Shell` which has hyperlink support

File: src/ops/view.rs [hunk 2]
```diff
@@ -12,6 +15,150 @@ use cargo::{
...
+                format!("https://docs.rs/{name}/{version}")
+            } else {
+                format!("https://docs.rs/{name}")
+            }),
+            Hyperlink::Keyword(keyword) => {
+                Cow::Owned(format!("https://crates.io/keywords/{keyword}"))
+            }
+            Hyperlink::Intrinsic(link) => Cow::Borrowed(*link),
+            Hyperlink::External(link) => Cow::Borrowed(*link),
+        }
```
Comment (epage) [inline src/ops/view.rs]: Why is knowledge of each type of link centralized here?  And whats the difference between Intrinsic and External?

File: src/ops/view.rs [hunk 3]
```diff
@@ -94,31 +265,36 @@ pub(super) fn pretty_view(
...
+                Hyperlink::Intrinsic(link)
+            } else {
+                Hyperlink::External(link)
+            }
         })
-    }) {
+        .or_else(|| {
+            is_crate_from_crates_io
+                .then(|| Hyperlink::Documentation(package_id.name(), Some(package_id.version())))
+        })
```
Comment (epage) [inline src/ops/view.rs]: A good number of terminals automatically turn URLs into links.  Do we need this?

Comment (epage) [reply src/ops/view.rs]: (ditto for other External links)

File: src/ops/view.rs [hunk 4]
```diff
@@ -28,12 +175,36 @@ pub(super) fn pretty_view(
...
+
+    if is_crate_from_crates_io {
+        write!(
+            stdout,
+            "{header}{}{header:#}",
+            Hyperlink::Crate(package_id.name(), /* Version */ None)
+        )?;
+    } else {
+        write!(stdout, "{header}{}{header:#}", package_id.name())?;
+    }
```
Comment (epage) [inline src/ops/view.rs:208]: This feels redundant with the crates.io link

File: src/ops/view.rs [hunk 5]
```diff
@@ -272,6 +448,8 @@ fn print_deps(
             FeatureStatus::EnabledByUser | FeatureStatus::Enabled => enabled,
             FeatureStatus::Disabled => disabled,
         };
+
+        // TODO: Show the link.
```
Comment (epage) [inline src/ops/view.rs:484]: This TODO should be resolved

File: src/ops/view.rs [hunk 6]
```diff
@@ -265,6 +281,22 @@ pub(super) fn pretty_view(
...
+                    .categories
+                    .iter()
+                    .map(|category| Hyperlink::Category(category).to_string())
+                    .collect::<Vec<_>>()
+                    .join(" ")
+            )?;
+        } else {
+            write!(stdout, " {}", metadata.categories.join(" "))?;
+        }
+    }
```
Comment (epage) [inline src/ops/view.rs:299]: This should be a separate PR (ideally it would have started as an Issue) as this needs more design conversation (and tests)

File: src/ops/view.rs (hunk 5 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~
I had resolved this on the next commit~

Please review again, thanks a lot😉~

File: src/ops/view.rs (hunk 4 above)
Comment (wangziling) [reply src/ops/view.rs:249]: Yes, I will remove the redundant one.

i.e. No need to make the `crate name` as a hyperlink 😉.

File: src/ops/view.rs (hunk 3 above)
Comment (wangziling) [reply src/ops/view.rs]: Yeah~

Hmmm... Okay~ I will sunset the transformation about the `already-be-link-format` contents.

File: src/ops/view.rs (hunk 2 above)
Comment (wangziling) [reply src/ops/view.rs]: Hmmm...

The `Intrinsic` means for the links which refer to the `crate.io` or `docs.rs` sites.

The `External` is the opposite against the `Intrinsic`.

These two are both be thought as a `newType` to identify the links.

File: src/ops/view.rs (hunk 1 above)
Comment (wangziling) [reply src/ops/view.rs:20]: Okay~ Give me some seconds to investigate~

File: src/ops/view.rs (hunk 6 above)
Comment (wangziling) [reply src/ops/view.rs:340]: Okay~ I will revert it~

Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had reverted it 😉~

File: src/ops/view.rs (hunk 4 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had sunset the hyperlink transformation about this one.
And sunset the transformation about the crate name(It will not be a link):
image

File: src/ops/view.rs (hunk 3 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had sunset the transformations about the already-be-link-format contents.

E.g. `documentation`, `crates.io`, `repository` and the `homepage` properties.

File: src/ops/view.rs (hunk 5 above)
Comment (wangziling) [reply src/ops/view.rs]: The result:
Only when the dependency version had reached out the `patch` level, it will be lead to the `crates.io/{its_name}/{its_version}` page, otherwise it wll be lead to the `crates.io/crates/{its_name}` page(Without exact version).

image

File: src/ops/view.rs (hunk 1 above)
Comment (wangziling) [reply src/ops/view.rs:20]: Hi Ed~

Yeah, the `GlobalContext` had owned the hyperlink supports, e.g., `out_hyperlink` and `err_hyperlink`, but it can only let the URL itself be a hyperlink, cannot accept the customized text(I will use the HTML/Javascript snippets as the examples):

`GlobalContext` only supports: `<a href="${url}">${url}</a>`.
The `Hyperlink` struct here can supports: `<a href="${url}">${customized_text}</a>`.

Comment (wangziling) [reply src/ops/view.rs:22]: Hi Ed~

I had used the `GlobalContext` to determine that the current terminal or `GlobalContext.shell` supports hyperlink or not.

Comment (wangziling) [reply src/ops/view.rs:22]: Hi Ed, I got what you mean.

https://github.com/rust-lang/cargo/blob/a8d72c675ee52dd57f0d8f2bae6655913c15b2fb/src/cargo/core/compiler/job_queue/mod.rs#L832

I'll change the implementation by following the codes above.

File: src/ops/view.rs (hunk 2 above)
Comment (epage) [reply src/ops/view.rs]: But why is it centralized?  I feel like this adds unnecessary extra layers.  Look up in Cargo how we are using `out_hyperlink` to see how we do links normally.

File: src/ops/view.rs (hunk 1 above)
Comment (epage) [reply src/ops/view.rs:22]: Why are you trying to construct it directly, rather than through `out_hyperlink`?  What is the immutable/mutable problem?

File: Cargo.toml [hunk 7]
```diff
@@ -20,7 +20,6 @@ crates-io = "0.40.0"
 pathdiff = "0.2.1"
 semver = "1.0.22"
 
-
```
Comment (epage) [inline Cargo.toml:23]: Stray newline left in

File: src/ops/view.rs [hunk 8]
```diff
@@ -95,12 +294,10 @@ pub(super) fn pretty_view(
...
-            format!(
-                "https://docs.rs/{name}/{version}",
-                name = package_id.name(),
-                version = package_id.version()
-            )
+        is_crate_from_crates_io.then(|| {
+            Hyperlink::Documentation(package_id.name(), Some(package_id.version()))
+                .as_link_lossy()
+                .into_owned()
         })
```
Comment (epage) [inline src/ops/view.rs:301]: Why is this still here?

File: src/ops/view.rs [hunk 9]
```diff
@@ -112,12 +309,11 @@ pub(super) fn pretty_view(
...
     // Only print the crates.io link if the package is from crates.io.
-    if summary.source_id().is_crates_io() {
+    if is_crate_from_crates_io {
         writeln!(
             stdout,
-            "{header}crates.io:{header:#} https://crates.io/crates/{}/{}",
-            package_id.name(),
-            package_id.version()
+            "{header}crates.io:{header:#} {}",
+            Hyperlink::Crate(package_id.name(), Some(package_id.version())).as_link_lossy()
```
Comment (epage) [inline src/ops/view.rs:316]: Why is this still here?

File: src/ops/view.rs (hunk 1 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

Thanks for your patience sincerely.

I'm just wondering not to change the existed `fn`s parameters, such as `print_deps` and `print_deps`...

My bad. I was stuck into my own thoughts.

File: src/ops/view.rs (hunk 2 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had removed these codes~

File: src/ops/view.rs (hunk 8 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had removed these codes.

File: src/ops/view.rs (hunk 9 above)
Comment (wangziling) [reply src/ops/view.rs]: Hi Ed~

I had removed these codes.

File: src/ops/view.rs [hunk 10]
```diff
@@ -272,13 +285,41 @@ fn print_deps(
...
-            "{style}{}{}{}{style:#}",
-            dependency.package_name(),
-            req,
-            source
-        )?;
+        let message = {
+            let name = dependency.package_name();
+            if dependency.source_id().is_crates_io() {
+                let link = match dependency.version_req() {
+                    cargo::util::OptVersionReq::Any => format!("https://crates.io/crates/{name}"),
```
Comment (epage) [inline src/ops/view.rs:292]: What is the right link for dependencies?
- crates.io? this command replaces a lot of use of it
- docs.rs?
- None, like package names?

Comment (epage) [reply src/ops/view.rs]: @hi-rustin thoughts?

File: src/ops/view.rs [hunk 11]
```diff
@@ -272,13 +285,41 @@ fn print_deps(
...
+                    cargo::util::OptVersionReq::Req(version_req)
+                    | cargo::util::OptVersionReq::Precise(_, version_req)
+                    | cargo::util::OptVersionReq::Locked(_, version_req) => {
+                        // Only when the version had reached to the `patch` level will it be valid to the `crates.io` registry.
```
[the remark this answers is not in the document]
Comment (epage) [reply src/ops/view.rs]: @hi-rustin thoughts?

**Comment by wangziling** [comment]
Hi @hi-rustin~ Could you review this PR and give me some advice?

Thanks a lot😉~

File: src/ops/view.rs (hunk 10 above)
Comment (hi-rustin) [reply src/ops/view.rs]: I'm not sure why I would need to check the docs.rs of a package using this command. Usually, I would just jump into the source using my IDE. If I do need to check the docs, I would search for them directly.

As for the crates.io link, in some scenarios, it may be possible to use it to see the popularity and maturity of a particular dependency or a number of dependencies.

And I checked the `npm info`, they don't include any links to dependencies.

So I prefer we only include the crates.io for them. The docs.rs link seems too warm for me.

Comment (epage) [reply src/ops/view.rs:292]: Should we just skip links for dependencies at this time?  There isn't one single answer for what to link to and, in theory, a "next step" is to instead run `cargo info` on the dependency.

Comment (hi-rustin) [reply src/ops/view.rs]: I agree that we can skip it for now.

**Review by hi-rustin — COMMENTED** [review]
Looks good to me. Thanks! 👍

Could you please squash your commits into one? Then we can merge it.

**Comment by wangziling** [comment]
> Looks good to me. Thanks! 👍
> 
> Could you please squash your commits into one? Then we can merge it.

Hi @hi-rustin~

Squashed 😉~

**Review by hi-rustin — APPROVED** [review]
Looks good to me. Thanks! 👍

**Comment by wangziling** [comment]
> Looks good to me. Thanks! 👍

You are welcome~~~

Thanks everyone's patience and advice😉~
